File: app/src/main/java/com/example/cinemates/model/local/dao/MovieDao.java
package com.example.cinemates.model.local.dao;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;

import com.example.cinemates.model.data.Movie;
import com.example.cinemates.model.data.PersonalStatus;

import java.util.List;

import io.reactivex.rxjava3.core.Observable;

/**
 * @author Antonio Di Nuzzo
 * Created 26/07/2022 at 07:36
 */
@Dao
public interface MovieDao {
    @Query("SELECT * FROM movie WHERE favorite = 1")
    Observable<List<Movie>> getAllFavorite();

    @Query("SELECT * FROM movie WHERE personalStatus = :status ORDER BY runtime ASC")
    Observable<List<Movie>> getAllWithStatus(PersonalStatus status);

    @Query("SELECT * FROM movie WHERE id = :id ")
    Movie retrieveMovie(Integer id);

    @Query("SELECT SUM(runtime)  FROM movie WHERE personalStatus =:status")
    long sumRuntimeAllWatchedMovies(PersonalStatus status);

    @Query("SELECT COUNT(id) FROM movie WHERE personalStatus = :status")
    long getMovieCountByStatus(PersonalStatus status);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertAllMovies(Movie... movies);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insert(Movie movie);


    @Update(onConflict = OnConflictStrategy.REPLACE)
    void update(Movie movie);

    @Delete
    void delete(Movie movie);
}

File: app/src/main/java/com/example/cinemates/model/local/dao/MovieDao.kt
package com.example.cinemates.model.local.dao

import androidx.room.*
import com.example.cinemates.model.entities.Movie
import kotlinx.coroutines.flow.Flow

/**
 * @author Antonio Di Nuzzo
 * Created 26/07/2022 at 07:36
 */
@Dao
interface MovieDao {
    @Query("SELECT * FROM movie")
    fun getMovies(): Flow<List<Movie>>

    @Query("SELECT * FROM movie WHERE id=:id")
    fun getMovie(id: Int): Movie?

    @Query("SELECT * FROM movie WHERE personalStatus=1")
    fun getToSeeMovies(): Flow<List<Movie>>

    @Query("SELECT * FROM movie WHERE personalStatus=0")
    fun getSeenMovies(): Flow<List<Movie>>

    @Query("SELECT * FROM movie WHERE favorite=1")
    fun getFavoriteMovies(): Flow<List<Movie>>

    @Query("SELECT EXISTS(SELECT * FROM movie WHERE id = :id AND favorite = 1)")
    fun isFavorite(id: Int): Boolean

    @Query("SELECT EXISTS(SELECT * FROM movie WHERE id = :id AND personalStatus = 1 )")
    fun isToSee(id: Int): Boolean

    @Query("SELECT EXISTS(SELECT * FROM movie WHERE id = :id AND personalStatus = 0)")
    fun isSeen(id: Int): Boolean

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(movie: Movie)

    @Update
    suspend fun update(movie: Movie)

    @Delete
    suspend fun delete(movie: Movie)
}